server/data/restClient.ts

```diff
@@ -90,31 +73,59 @@ export default class RestClient {
     data = {},
     raw = false,
   }: PostRequest = {}): Promise<unknown> {
-    this.logRequest('POST', path)
+    return this.processRequest({ path, headers, responseType, data, raw, method: 'post' })
+  }
 
-    const request = superagent.post(`${this.apiUrl()}${path}`)
+  async put({
+    path = null,
+    headers = {},
+    responseType = '',
+    data = {},
+    raw = false,
+  }: PutRequest = {}): Promise<unknown> {
+    return this.processRequest({ path, headers, responseType, data, raw, method: 'put' })
+  }
+
+  private async processRequest({
+    path = null,
+    headers = {},
+    responseType = '',
+    query = undefined,
+    data = undefined,
+    raw = false,
+    method = undefined,
+  } = {}): Promise<unknown> {
+    this.logRequest(method.toUpperCase(), path)
+
+    const request = superagent[method](`${this.apiUrl()}${path}`)
     request
-      .send(data)
       .agent(this.agent)
-      .retry(2, err => {
-        if (err) logger.info(`Retry handler found API error with ${err.code} ${err.message}`)
+      .retry(2, (err: ResponseError): void => {
+        if (err) logger.info(`Retry handler found API error with ${err.status} ${err.message}`)
         return undefined // retry handler only for logging retries, not to influence retry logic
       })
       .set(headers)
       .responseType(responseType)
       .timeout(this.timeoutConfig())
 
+    if (data) {
+      request.send(data)
+    }
+    if (query) {
+      request.query(query)
+    }
+
     if (this.hmppsToken) {
       request.auth(this.hmppsToken, { type: 'bearer' })
     } else if (this.slmToken) {
       request.set('Create-Barcode-Token', this.slmToken)
     }
```

Comment:
I like this a lot (a single method to invoke the actual http request), and right here is where I'm going to add basic auth support in my next PR 👍 😄